website/docs/metaprogramming-plugins.md

```diff
@@ -0,0 +1,124 @@
+---
+id: metaprogramming-plugins
+title: Metaprogramming plugins
+---
+
+> **Note**: The plugin system has many drawbacks and is not considered a stable
+> feature. Expect its API to change in the future.
+
+In Ruby, it is possible to dynamically define methods. Consider the following
+example:
+
+```ruby
+# metaprogramming.rb
+# typed: true
+
+class MetaProgramming
+  def self.macro(name)
+    define_method("#{name}_time") { name }
+  end
+
+  macro(:bed)
+  macro(:fun)
+end
+
+hello = MetaProgramming.new
+hello.bed_time # error: Method `bed_time` does not exist on MetaProgramming
+hello.fun_time # error: Method `fun_time` does not exist on MetaProgramming
+```
+
+Sorbet rejects this script since currently, Sorbet does not try to understand
+calls to `define_method` or other similar metaprogramming facilities.
+
+Sorbet has a plugin system for handling metaprogramming patterns similar to the
+one shown in this example. Specifically, we can write a plugin to teach Sorbet
+what the method `macro` does.
+
+Here is a plugin for `macro` in the example above:
+
+```ruby
+# macro_plugin.rb
+
+# Sorbet calls this plugin with command line arguments similar to the following:
+# --class MetaProgramming --method macro --source macro(:bed)
+# we only care about the source here, so we use ARGV[5]
+source = ARGV[5]
+/macro\(:(.*)\)/.match(source) do |match_data|
+  puts "def #{match_data[1]}_time; end"
+end
+```
+
+We then use a YAML file to tell Sorbet about this plugin.
+
+```yaml
+# triggers.yaml
+
+ruby_extra_args:
+  # These options are forwarded to Ruby
+  - '--disable-gems' # This option speeds up Ruby boot time. Use it if you don't need gems
+triggers:
+  macro: macro_plugin.rb # This tells Sorbet to run macro.rb when it sees a call to `macro`
+```
+
+Let's run Sorbet on our initial example again, this time with our new plugin
+enabled:
+
+```shell
+❯ srb tc --dsl-plugins triggers.yaml metaprogramming.rb
+No errors! Great job.
+```
+
+Sorbet executed our plugin and took into consideration its output. Method
+definitions generated by our plugin fixed the error we saw initially.
+
+## Debugging plugins
+
+We can ask Sorbet to print out the output of all plugin calls using
+`--print plugin-generated-code`:
+
+```shell
+❯ srb tc --print plugin-generated-code --dsl-plugins triggers.yaml metaprogramming.rb
+# Path: "metaprogramming.rb//plugin-generated|0":
+class MetaProgramming;
+def bed_time; end
+end;
+# Path: "metaprogramming.rb//plugin-generated|1":
+class MetaProgramming;
+def fun_time; end
+end;
+No errors! Great job.
+```
+
+You can also add debug outputs by printing to `$stderr` within your plugin.
```

Comment:
```suggestion
Anything printed to `$stderr` within a plugin will show up in the terminal when
`srb tc` runs.
```

reworded to avoid "you" and "your"

Reply:
The extra info about where it shows up is also nice.